src/gromacs/nblib/topology.cpp

```diff
@@ -96,20 +107,30 @@ std::vector<real> TopologyBuilder::extractQuantity(Extractor extractor)
             {
                 std::string atomTypeName = std::get<1>(atomTuple);
 
+                if(atomTypeIndex.count(atomTypeName) == 0) {
+                    atomTypeIndex[atomTypeName] = numAtomTypes;
+                    numAtomTypes++;
+                }
+
                 Atom &atomType = molecule.atomTypes_[atomTypeName];
-                ret.push_back(extractor(atomType));
+
+                topology_.masses_.push_back(atomType.mass());
+                topology_.charges_.push_back(atomType.charge());
+                topology_.nonbondedParameters_.push_back(atomType.c6());
+                topology_.nonbondedParameters_.push_back(atomType.c12());
+
+                SET_CGINFO_HAS_VDW(topology_.atomInfoAllVdw_[numAtoms]);
+                SET_CGINFO_HAS_Q(topology_.atomInfoAllVdw_[numAtoms]);
+                numAtoms++;
             }
         }
     }
-
-    return ret;
 }
 
 Topology TopologyBuilder::buildTopology()
 {
-    topology_.excls = createExclusionsList();
-    topology_.masses = extractQuantity([](const Atom &atomType){ return atomType.mass(); });
-    topology_.charges = extractQuantity([](const Atom &atomType){ return atomType.charge(); });
+    topology_.excls_ = createExclusionsList();
+    extractAllProperties();
```

Comment:
Now these two different methods have different interfaces. When someone comes along to extend the topology class how should they pick whether their setup functionality has a return or is void?